File: framework/cluster_manager/jvm/main/java/com/webank/eggroll/framework/clustermanager/client/ClusterManagerClient.java
/*
 * Copyright (c) 2019 - now, Eggroll Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *
 */

package com.webank.eggroll.framework.clustermanager.client;

import com.webank.eggroll.clustermanager.constant.MetadataCommands;
import com.webank.eggroll.core.command.CommandClient;
import com.webank.eggroll.core.command.CommandURI;
import com.webank.eggroll.core.constant.ClusterManagerConfKeys;
import com.webank.eggroll.core.constant.SerdesTypes;
import com.webank.eggroll.core.datastructure.RpcMessage;
import com.webank.eggroll.core.meta.ErEndpoint;
import com.webank.eggroll.core.meta.ErPartition;
import com.webank.eggroll.core.meta.ErServerCluster;
import com.webank.eggroll.core.meta.ErServerNode;
import com.webank.eggroll.core.meta.ErStore;
import com.webank.eggroll.core.meta.ErStoreLocator;
import com.webank.eggroll.core.session.DefaultErConf;

public class ClusterManagerClient {
  private ErEndpoint clusterManagerEndpoint;
  private CommandClient commandClient;
  private static final ErPartition[] EMPTY_PARTITION_ARRAY = new ErPartition[0];

  public ClusterManagerClient(String serverHost, int serverPort) {
    this.clusterManagerEndpoint = new ErEndpoint(serverHost, serverPort);
    this.commandClient = new CommandClient();
  }

  public ClusterManagerClient() {
    this(DefaultErConf.getString(
            ClusterManagerConfKeys.CONFKEY_CLUSTER_MANAGER_HOST(), null),
        DefaultErConf.getInt(
            ClusterManagerConfKeys.CONFKEY_CLUSTER_MANAGER_PORT(), -1));
  }

  public ErServerNode getServerNode(ErServerNode input) {
    return doSyncRequestInternal(input, ErServerNode.class, MetadataCommands.GET_SERVER_NODE());
  }

  public ErServerCluster getServerNodes(ErServerNode input) {
    return doSyncRequestInternal(input, ErServerCluster.class, MetadataCommands.GET_SERVER_NODES());
  }

  public ErServerNode getOrCreateServerNode(ErServerNode input) {
    return doSyncRequestInternal(input, ErServerNode.class, MetadataCommands.GET_OR_CREATE_SERVER_NODE());
  }

  public ErServerNode createOrUpdateServerNode(ErServerNode input) {
    return doSyncRequestInternal(input, ErServerNode.class, MetadataCommands.CREATE_OR_UPDATE_SERVER_NODE());
  }

  public ErStore getStore(ErStoreLocator input) {
    return getStore(new ErStore(input, EMPTY_PARTITION_ARRAY));
  }

  public ErStore getStore(ErStore input) {
    return doSyncRequestInternal(input, ErStore.class, MetadataCommands.GET_STORE());
  }

  public ErStore getOrCreateStore(ErStoreLocator input) {
    return getOrCreateStore(new ErStore(input, EMPTY_PARTITION_ARRAY));
  }

  public ErStore getOrCreateStore(ErStore input) {
    return doSyncRequestInternal(input, ErStore.class, MetadataCommands.GET_OR_CREATE_STORE());
  }

  public ErStore deleteStore(ErStore input) {
    return doSyncRequestInternal(input, ErStore.class, MetadataCommands.DELETE_STORE());
  }

  public ErStore deleteStore(ErStoreLocator input) {
    return deleteStore(new ErStore(input, EMPTY_PARTITION_ARRAY));
  }

  private <T> T doSyncRequestInternal(RpcMessage input, Class<T> outputClass, CommandURI commandURI) {
    return commandClient.simpleSyncSend(input, outputClass, clusterManagerEndpoint, commandURI, SerdesTypes.PROTOBUF());
  }
}

File: jvm/core/main/scala/com/webank/eggroll/core/client/ClusterManagerClient.scala
/*
 * Copyright (c) 2019 - now, Eggroll Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 *
 */

package com.webank.eggroll.core.client

import java.util.concurrent.ConcurrentHashMap

import com.webank.eggroll.core.command.CommandClient
import com.webank.eggroll.core.constant.{ClusterManagerConfKeys, MetadataCommands, SessionCommands}
import com.webank.eggroll.core.meta._
import com.webank.eggroll.core.session.StaticErConf

import scala.collection.JavaConverters._

class ClusterManagerClient(val endpoint: ErEndpoint) {
  if (endpoint == null || !endpoint.isValid) {
    throw new IllegalArgumentException(s"failed to create ClusterManagerClient for endpoint: ${endpoint}")
  }

  private val cc = new CommandClient(endpoint)
  private val EMPTY_PARTITION_ARRAY = Array[ErPartition]();

  def this(serverHost: String, serverPort: Int) {
    this(ErEndpoint(serverHost, serverPort));
  }

  // TODO:2: priority property getter
  def this(options: Map[String, String]) {
    this(options.getOrElse(ClusterManagerConfKeys.CONFKEY_CLUSTER_MANAGER_HOST,
      StaticErConf.getString(ClusterManagerConfKeys.CONFKEY_CLUSTER_MANAGER_HOST,
        "")),
      options.getOrElse(ClusterManagerConfKeys.CONFKEY_CLUSTER_MANAGER_PORT,
        StaticErConf.getString(ClusterManagerConfKeys.CONFKEY_CLUSTER_MANAGER_PORT,
          "-1")).toInt)
  }

  def this() {
    this(StaticErConf.getString(
      ClusterManagerConfKeys.CONFKEY_CLUSTER_MANAGER_HOST, ""),
      StaticErConf.getInt(
        ClusterManagerConfKeys.CONFKEY_CLUSTER_MANAGER_PORT, -1))
  }

  def getServerNode(input:ErServerNode):ErServerNode = cc.call[ErServerNode](MetadataCommands.GET_SERVER_NODE, input)

  def getServerNodes(input:ErServerNode):ErServerCluster = cc.call[ErServerCluster](MetadataCommands.GET_SERVER_NODES, input)

  def getOrCreateServerNode(input:ErServerNode):ErServerNode =
    cc.call[ErServerNode](MetadataCommands.GET_OR_CREATE_SERVER_NODE, input)

  def createOrUpdateServerNode(input: ErServerNode):ErServerNode =
    cc.call[ErServerNode](MetadataCommands.CREATE_OR_UPDATE_SERVER_NODE, input)

  def getStore(input:ErStoreLocator):ErStore =
    getStore(ErStore(input, EMPTY_PARTITION_ARRAY, Map[String, String]().asJava))

  def getStore(input: ErStore): ErStore = cc.call[ErStore](MetadataCommands.GET_STORE, input)

  def getOrCreateStore(input: ErStoreLocator): ErStore =
    getOrCreateStore(new ErStore(input, EMPTY_PARTITION_ARRAY, new ConcurrentHashMap[String, String]))

  def getOrCreateStore(input: ErStore): ErStore = cc.call[ErStore](MetadataCommands.GET_OR_CREATE_STORE, input)

  def deleteStore(input: ErStore): ErStore = cc.call[ErStore](MetadataCommands.DELETE_STORE, input)

  def deleteStore(input: ErStoreLocator): ErStore =
    deleteStore(new ErStore(input, EMPTY_PARTITION_ARRAY, new ConcurrentHashMap[String, String]))

  def getStoreFromNamespace(input: ErStore): ErStoreList = cc.call[ErStoreList](MetadataCommands.GET_STORE_FROM_NAMESPACE, input)

  def getStoreFromNamespace(input: ErStoreLocator): ErStoreList = {
    getStoreFromNamespace(new ErStore(input, EMPTY_PARTITION_ARRAY, new ConcurrentHashMap[String, String]))
  }
  def getOrCreateSession(sessionMeta: ErSessionMeta): ErSessionMeta =
    cc.call[ErSessionMeta](SessionCommands.getOrCreateSession, sessionMeta)

  def getSession(sessionMeta: ErSessionMeta): ErSessionMeta =
    cc.call[ErSessionMeta](SessionCommands.getSession, sessionMeta)

  def stopSession(sessionMeta: ErSessionMeta): ErSessionMeta =
    cc.call[ErSessionMeta](SessionCommands.stopSession, sessionMeta)

  def killSession(sessionMeta: ErSessionMeta): ErSessionMeta =
    cc.call[ErSessionMeta](SessionCommands.killSession, sessionMeta)

  def killAllSessions(): Unit =
    cc.call[ErSessionMeta](SessionCommands.killAllSessions, ErSessionMeta())

  def registerSession(sessionMeta: ErSessionMeta): ErSessionMeta =
    cc.call[ErSessionMeta](SessionCommands.registerSession, sessionMeta)

  // todo:0: heartbeat with process pid
  def heartbeat(processor: ErProcessor): ErProcessor =
    cc.call[ErProcessor](SessionCommands.heartbeat, processor)
}
